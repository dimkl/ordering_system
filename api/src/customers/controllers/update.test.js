/**
 * @integration-test true
 * @data-factory true
 */
describe("PATCH /customers/:customer_id", () => {
  let knex;
  beforeAll(() => knex = require('../../shared/setupModels')());
  beforeEach(() => knex.raw('truncate customers, orders cascade;'));
  afterAll(() => knex.destroy());

  it("updates and returns updated customer", async () => {
    const customer = await DataFactory.createCustomer();

    const response = await request.patch(`/customers/${customer.uuid}`)
      .send({
        "first_name": "Dimitris+2",
        "last_name": "Klouvas+2",
        "email": "dimitris.klouvas+2@gmail.com",
        "password": "123456"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchObject({
      "email": "dimitris.klouvas+2@gmail.com",
      first_name: "Dimitris+2",
      last_name: "Klouvas+2",
      id: customer.id,
      uuid: customer.uuid,
      created_at: customer.created_at.toISOString(),
      updated_at: expect.any(String)
    });
    const unixUpdatedAt = new Date(response.body.updated_at).getTime();
    expect(unixUpdatedAt).toBeGreaterThan(customer.updated_at.getTime());
  });

  it("does not have required properties", async () => {
    const customer = await DataFactory.createCustomer();

    const response = await request.patch(`/customers/${customer.uuid}`)
      .send({})
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
    });
  });

  it("throws validation error for unique email", async () => {
    const existingCustomer = await DataFactory.createCustomer();
    const newCustomer = await DataFactory.createCustomer({email: 'customer-update@example.com'});

    const response = await request.patch(`/customers/${newCustomer.uuid}`)
      .send({
        "first_name": "Dimitris",
        "last_name": "Klouvas",
        "email": existingCustomer.email,
        "password": "1234"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(422);
    expect(response.body).toMatchSnapshot();
  });

  it("omits additional properties", async () => {
    const customer = await DataFactory.createCustomer();

    const response = await request.patch(`/customers/${customer.uuid}`)
      .send({ created_at: "1680046371850", first_name: "aloha" })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.created_at).not.toEqual("1680046371850");
    expect(response.body.first_name).toEqual("aloha");
  });
});