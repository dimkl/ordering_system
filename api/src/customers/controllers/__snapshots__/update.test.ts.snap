// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PATCH /customers/:customer_id does not have required properties 1`] = `
{
  "created_at": Any<String>,
  "email": "dimitris.klouvas@gmail.com",
  "first_name": "Dimitris",
  "id": Any<String>,
  "last_name": "Klouvas",
  "updated_at": Any<String>,
}
`;

exports[`PATCH /customers/:customer_id throws validation error for unique email 1`] = `
{
  "message": "email already exists!",
}
`;
